### PR TITLE
[#1083] Try block syntax

### PR DESCRIPTION
**Pull request by andreytkachenko**
The `try` block syntax

### CODE REVIEW

File: crates/ra_hir/src/ty/infer.rs [hunk 1]
```diff
@@ -946,6 +946,12 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 then_ty
             }
             Expr::Block { statements, tail } => self.infer_block(statements, *tail, expected),
+            Expr::TryBlock { body } => {
+                let _inner = self.infer_expr(*body, expected);
+                
+                // FIXME should be std::result::Result<{inner}, _>
+                Ty::Unknown
```
Comment (andreytkachenko) [inline crates/ra_hir/src/ty/infer.rs]: How can I do that? Is any guide how can I pull std::result::Result type here and return with substitution of generic type?
I have tried different attempts to make it correct, but it doesn't works:
```rust
                let res = self.resolver.resolve_name(self.db, &Name::new("::std::result::Result".into()))
                    .take_types()
                    .unwrap();

                let res_enum = match res {
                    Resolution::Def(ModuleDef::Enum(val)) => val,
                    _ => unreachable!()
                };

                Ty::apply(TypeCtor::Adt(AdtDef::Enum(res_enum)), Substs(Arc::new([inner, Ty::Unknown])))
```

Comment (matklad) [reply crates/ra_hir/src/ty/infer.rs]: I think this needs to be more involved and go via the `Try` trait, and we don't quite yet have all the ground-work in place to support it.

Comment (andreytkachenko) [reply crates/ra_hir/src/ty/infer.rs]: @matklad is any issue to track such ground-work?

Comment (matklad) [reply crates/ra_hir/src/ty/infer.rs]: See https://github.com/rust-analyzer/rust-analyzer/issues/909#issuecomment-499631125

**Pull request merged**